consensus/src/transaction_shuffler/use_case_aware/delayed_queue.rs

```diff
@@ -0,0 +1,503 @@
+        }
+        (account_delay_key, address)
+    }
+
+    fn update_try_delay_till(&mut self, try_delay_till: OutputIdx) {
+        self.try_delay_till = try_delay_till;
+    }
+
+    fn add_account<Txn>(&mut self, address: AccountAddress, account: &Account<Txn>)
+    where
+        Txn: UseCaseAwareTransaction,
+    {
+        let account_delay_key = account.delay_key();
+        self.account_by_delay
+            .strict_insert(account_delay_key, address);
+        let (_, head_address) = self.account_by_delay.first_key_value().unwrap();
+        if head_address == &address {
+            self.input_idx = account_delay_key.input_idx;
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct DelayedQueue<Txn> {
+    accounts: HashMap<AccountAddress, Account<Txn>>,
+    use_cases: HashMap<UseCaseKey, UseCase>,
+
+    /// empty accounts (those w/o known delayed txns), kept to track the delay
+    account_placeholders_by_delay: BTreeMap<DelayKey, AccountAddress>,
+    /// empty use cases (those w/o known delayed txns), kept to track the delay
+    use_case_placeholders_by_delay: BTreeMap<DelayKey, UseCaseKey>,
+
+    /// main delay queue, all use cases are non-empty of non-empty accounts
+    use_cases_by_delay: BTreeMap<DelayKey, UseCaseKey>,
+
+    /// externally set output index; when an item has try_delay_till <= output_idx, it's deemed ready
+    output_idx: OutputIdx,
+
+    config: Config,
+}
+
+impl<Txn> DelayedQueue<Txn>
+where
+    Txn: UseCaseAwareTransaction,
+{
+    pub fn new(config: Config) -> Self {
+        Self {
+            accounts: HashMap::new(),
+            use_cases: HashMap::new(),
+
+            account_placeholders_by_delay: BTreeMap::new(),
+            use_case_placeholders_by_delay: BTreeMap::new(),
+
+            use_cases_by_delay: BTreeMap::new(),
+
+            output_idx: 0,
+
+            config,
+        }
+    }
+
+    /// Remove stale (empty use cases and accounts with try_delay_till <= self.output_idx) placeholders.
+    fn drain_placeholders(&mut self) {
+        let least_to_keep = DelayKey::new(self.output_idx + 1, 0);
+
+        let remaining_use_case_placeholders = self
+            .use_case_placeholders_by_delay
+            .split_off(&least_to_keep);
+        let remaining_account_placeholders =
+            self.account_placeholders_by_delay.split_off(&least_to_keep);
+
+        self.use_case_placeholders_by_delay
+            .iter()
+            .for_each(|(_delay_key, use_case_key)| self.use_cases.strict_remove(use_case_key));
+        self.account_placeholders_by_delay
+            .iter()
+            .for_each(|(_delay_key, address)| self.accounts.strict_remove(address));
+
+        self.use_case_placeholders_by_delay = remaining_use_case_placeholders;
+        self.account_placeholders_by_delay = remaining_account_placeholders;
+    }
+
+    pub fn bump_output_idx(&mut self, output_idx: OutputIdx) {
+        self.output_idx = output_idx;
```

Comment:
can we just assert that new output_idx is larger than self.output_idx?


also, we increment this by 1, but this structure doesn't depend on that, correct? (i.e. increment by 3 works just fine?)

Reply:
> can we just assert that new output_idx is larger than self.output_idx?
sure
> also, we increment this by 1, but this structure doesn't depend on that, correct? (i.e. increment by 3 works just fine?)

I think so.. it shouldn't break even if we make output_idx smaller -_,-